test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,1657 @@
+        );
+
+        vm.prank(address(ethPaymentTerminal));
+        dominantJuice_twoPledgers.didPay(didPayData_p2);
+
+        (uint256 pledger2W, uint256 totalW2) = dominantJuice_twoPledgers.exposed_getPledgerAndTotalWeights(pledger2);
+
+        // Multiply 0.99 * 100 to avoid decimals. Then divide by the factored out 100^5 at end to get final result.
+        uint256 timeFactor2 = 99 ** hourOfPledge2;
+        uint256 calculatedWeight2 = timeFactor2 * 100 ether / (100 ** 5);
+        console.log("calculatedWeight2", calculatedWeight2);
+
+        console.log("pledger2W", pledger2W);
+        console.log("totalW2", totalW2);
+        assertEq(pledger2W, totalW2 - totalW1);
+        assertGt(pledger1W, pledger2W);
+    }
+
+    // PASSING - Happy path
+    function test_didPay_multiplePledgers_campaignUpdates() public {
+        bonusDeposited();
+
+        vm.warp(3600); // warp 1 hour
+        pledge(pledger1, MIN_PLEDGE_AMOUNT);
+
+        vm.warp(86400); // warp 1 day
+        pledge(pledger2, MIN_PLEDGE_AMOUNT * 3);
+
+        vm.warp(604800); // warp 1 week
+        pledge(pledger3, MIN_PLEDGE_AMOUNT);
+
+        DominantJuice.FundingStatus memory status = dominantJuice.getCycleFundingStatus();
+
+        assertEq(MIN_PLEDGE_AMOUNT * 5, status.totalPledged);
+        uint256 percent = (100 * MIN_PLEDGE_AMOUNT * 5) / CYCLE_TARGET;
+        assertEq(percent, status.percentOfGoal);
+        assertEq(false, status.isTargetMet);
+    }
+
+    // PASSING - Three pledgers pledge same amount at different times.
+    function test_didPay_multiplePledgersSameAmountDiffTimes_weights() public {
+        (DominantJuiceTestHelper dominantJuice_multipledger, JBFundingCycle memory cycleData) =
+            dominantJuiceTesterSetup();
+
+        vm.warp(2 days + 1);
+        pledgeToTester(dominantJuice_multipledger, cycleData, pledger1, MIN_PLEDGE_AMOUNT);
+
+        (uint256 pledger1W,) = dominantJuice_multipledger.exposed_getPledgerAndTotalWeights(pledger1);
+
+        vm.warp(block.timestamp + 1 days);
+        pledgeToTester(dominantJuice_multipledger, cycleData, pledger2, MIN_PLEDGE_AMOUNT);
+
+        (uint256 pledger2W, uint256 totalW2) = dominantJuice_multipledger.exposed_getPledgerAndTotalWeights(pledger2);
+
+        vm.warp(block.timestamp + 1 weeks);
+        pledgeToTester(dominantJuice_multipledger, cycleData, pledger3, MIN_PLEDGE_AMOUNT);
+
+        (uint256 pledger3W, uint256 totalW3) = dominantJuice_multipledger.exposed_getPledgerAndTotalWeights(pledger3);
+
+        // Test weights with ABDKLibrary here once pow() value resutls are figured out?
+
+        assertLt(pledger2W, pledger1W); // pledger2W is less since pledged later
+        assertLt(pledger3W, pledger1W); // pledger3W is less since pledged much later
+        assertEq(pledger1W + pledger2W, totalW2); // 1st and 2nd pledger weights equal total weight
+        assertEq(pledger1W + pledger2W + pledger3W, totalW3); // All three pledges equal total weight
```

Comment:
You have some good multiple pledger tests here. Only comment is that I'm less worried about the weights as that's an internal calculation as a means to an end. I'm more worried that the value that the pledger can take out is correct, and unless i'm missing it, I'm not seeing you check those values. The outcome is more important than the internal calculations of the weights. You can also check the outcome against a calculator based on what you think it should be. If the outcomes are wrong, maybe it means that the weight calculations are wrong, but that's more of a debugging step than a test.

Reply:
Understood on the weight tests. There aren't any withdraw/balance checks during the didPay() tests. I assert the contract balance a bunch in the didRedeem() tests, as well as the pledger wallets a few times (I tried to only do this a few times, as I'm told that in unit tests you should only focus on state in the contract at hand, not state changes in other addresses or contracts.)